tests/functional/keystoneapi_controller_test.go

```diff
@@ -1109,6 +1110,112 @@ var _ = Describe("Keystone controller", func() {
 		})
 	})
 
+	When("When the fernet keys are created with fernetMaxActiveKeys as 3", func() {
```

Comment:
Both this test and the one below are pretty much similar, is there a way to reuse it?
(While waiting for an answer I'll try to figure it out, but it may take some time)

Reply:
instead of the separate test bellow, maybe it would be better to test that when updating the keystone parameter from 3 -> 6 we see that the config and the secret got additional keys? Like have a nested when update keys from 3->6 it ...

Reply:
The problem is that I still don't know how to simulate a change on the spec.
I start with this one: DeferCleanup(th.DeleteInstance, CreateKeystoneAPI(keystoneAPIName, GetKeystoneAPISpec(3)))
But even if I call this again with a different value on the It(), it doesn't seem to work.

Reply:
you could do it like this

```
	When("When the fernet keys are created", func() {
		BeforeEach(func() {
			DeferCleanup(
				k8sClient.Delete, ctx, CreateKeystoneMessageBusSecret(namespace, "rabbitmq-secret"))
			// with fernetMaxActiveKeys as 3
			DeferCleanup(th.DeleteInstance, CreateKeystoneAPI(keystoneAPIName, GetKeystoneAPISpec(3)))
			DeferCleanup(
				k8sClient.Delete, ctx, CreateKeystoneAPISecret(namespace, SecretName))
			DeferCleanup(infra.DeleteMemcached, infra.CreateMemcached(namespace, "memcached", memcachedSpec))
			DeferCleanup(
				mariadb.DeleteDBService,
				mariadb.CreateDBService(
					namespace,
					GetKeystoneAPI(keystoneAPIName).Spec.DatabaseInstance,
					corev1.ServiceSpec{
						Ports: []corev1.ServicePort{{Port: 3306}},
					},
				),
			)
			mariadb.SimulateMariaDBAccountCompleted(keystoneAccountName)
			mariadb.SimulateMariaDBDatabaseCompleted(keystoneDatabaseName)
			infra.SimulateTransportURLReady(types.NamespacedName{
				Name:      fmt.Sprintf("%s-keystone-transport", keystoneAPIName.Name),
				Namespace: namespace,
			})
			infra.SimulateMemcachedReady(types.NamespacedName{
				Name:      "memcached",
				Namespace: namespace,
			})
			th.SimulateJobSuccess(dbSyncJobName)
			th.SimulateJobSuccess(bootstrapJobName)
			th.SimulateDeploymentReplicaReady(deploymentName)
		})

		It("creates 3 keys", func() {
			secret := th.GetSecret(types.NamespacedName{Namespace: keystoneAPIName.Namespace, Name: "keystone"})
			Expect(secret).ToNot(BeNil())

			Eventually(func(g Gomega) {
				Expect(maps.Keys(secret.Data)).To(ContainElements("FernetKeys0", "FernetKeys1", "FernetKeys2"))
				Expect(maps.Keys(secret.Data)).ToNot(ContainElement("FernetKeys3"))

				for i := 0; i < 3; i++ {
					g.Expect(secret.Data["FernetKeys"+strconv.Itoa(i)]).NotTo(BeZero())
				}
			}, timeout, interval).Should(Succeed())
		})

		When("When the fernet keys get updated to 6", func() {
			BeforeEach(func() {
				keystone := GetKeystoneAPI(keystoneAPIName)
				Expect(*keystone.Spec.FernetMaxActiveKeys).To(Equal(int32(3)))

				keystone.Spec.FernetMaxActiveKeys = ptr.To(int32(6))

				_, err := controllerutil.CreateOrPatch(
					th.Ctx, th.K8sClient, keystone, func() error { return nil })
				Expect(err).ToNot(HaveOccurred())
			})
			It("creates 6 keys", func() {
				secret := th.GetSecret(types.NamespacedName{Namespace: keystoneAPIName.Namespace, Name: "keystone"})
				Expect(secret).ToNot(BeNil())

				Eventually(func(g Gomega) {
					Expect(maps.Keys(secret.Data)).To(ContainElements("FernetKeys0", "FernetKeys1", "FernetKeys2", "FernetKeys3", "FernetKeys4", "FernetKeys5"))
					Expect(maps.Keys(secret.Data)).ToNot(ContainElement("FernetKeys6"))

					for i := 0; i < 6; i++ {
						g.Expect(secret.Data["FernetKeys"+strconv.Itoa(i)]).NotTo(BeZero())
					}
				}, timeout, interval).Should(Succeed())
			})
		})
	})
```

when you run this it fails with

```
  [FAILED] Expected
      <[]string | len:5, cap:5>: ["FernetKeys2", "CredentialKeys0", "CredentialKeys1", "FernetKeys0", "FernetKeys1"]
  to contain elements
      <[]string | len:6, cap:6>: ["FernetKeys0", "FernetKeys1", "FernetKeys2", "FernetKeys3", "FernetKeys4", "FernetKeys5"]
  the missing elements were
      <[]string | len:3, cap:3>: ["FernetKeys3", "FernetKeys4", "FernetKeys5"]
  In [It] at: /home/mschuppe/src/github.com/openstack-k8s-operators/keystone-operator/tests/functional/keystoneapi_controller_test.go:1178 @ 10/25/24 12:34:45.138
```

so I think there might be a bug with the logic which adds additional keys?

Reply:
Thanks for the tip on updating the spec.
Regarding the test part:
```
				Expect(maps.Keys(secret.Data)).To(ContainElements("FernetKeys0", "FernetKeys1", "FernetKeys2"))
				Expect(maps.Keys(secret.Data)).ToNot(ContainElement("FernetKeys3"))
```
I prefer to make sure there's no leftovers, for example, if we decrease from 6 to 3 keys, if somehow we ended up in this scenario: FernetKeys0, 1, 2, 5; I wanna make sure ALL the 'FernetKeys%d' are what we expect, so I`m counting how many FK we have and that they are all what we expect. Other way to do this (That I know of) would be to get each FK and see if they are valid; to check that they are valid would be to iterate over a range of fernetMaxKeys, create a string like 'FernetKeys%d' and add it to a list.
I preferred just to count the FernetKeys keys on the secret.

Reply:
I am not saying you should use from above, but you could still do the same with that, like what I did that a key must not exist. I prefer that as it is easy to read. e.g. when you go from 6 to 3, you:
* expect these
Expect(maps.Keys(secret.Data)).To(ContainElements("FernetKeys0", "FernetKeys1", "FernetKeys2"))
* you expect that those are gone
 Expect(maps.Keys(secret.Data)).ToNot(ContainElements("FernetKeys3", "FernetKeys4", "FernetKeys5"))

What you probably want to enhance, e.g. when you go from 3 to 6 that the last key is the expected one. iiuc in this scenario, the old key[2] should become the key[5], right?